### PR TITLE
CORE-10381 Improving resilience of external event flow against transient exceptions

### PR DESCRIPTION
**Pull request by ben-millar**
This PR is a proposed fix to the issue raised in [CORE-10381](https://r3-cev.atlassian.net/browse/CORE-10381).
### Root cause:
When a flow hits a transient exception while handling some external event, it will cache that event and rollback the checkpoint *provided* that we think the event will succeed at some point in the future. Then, the next time we wakeup the flow, we retry this event and then clear the `inRetryState` flag.

What's happening here, however, is that some other event is coming in before we have a chance to replay the failed event and then clearing the `inRetryState` flag and the stored event.

### Fix
To fix this, we simply set a flag in the pre-processing step if the event we're handling is the one that triggered the retry. Only when this flag is true should we clear the `inRetryState` flag, and the event to be retried, in the post processing step.

Additionally, the `shouldRetry` logic has been updated. Previously this would re-send an external event only if we'd received back a `retry` response from the external worker, but this has been updated such that *all* non-error states should be retried *provided* that the resend window has elapsed. This handles situations where a `retry` response has not explicitly been sent, but we should still retry the event.

## Notes
Acceptance tests have been updated to reflect the new expected behaviour.

[CORE-10381]: https://r3-cev.atlassian.net/browse/CORE-10381?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/state/FlowStateManagerTest.kt [hunk 1]
```diff
@@ -49,7 +54,7 @@ class FlowStateManagerTest {
     }
 
     @Test
-    fun `Rollback successfully restores sessions to initial state adding additional sessions`() {
+    fun `Rollback successfully restores sessions to initial state after adding additional sessions`() {
```
Comment (ben-millar) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/state/FlowStateManagerTest.kt:52]: Finally fixed this typo @emilybowe 😁

File: components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ExternalEventAcceptanceTest.kt [hunk 2]
```diff
@@ -221,8 +221,9 @@ class ExternalEventAcceptanceTest : FlowServiceTestBase() {
 
     @Test
     fun `Receiving an event does not resend the external event unless a 'transient' error is received`() {
-
         given {
+            flowConfiguration(FlowConfig.EXTERNAL_EVENT_MESSAGE_RESEND_WINDOW, 500000L)
```
Comment (emilybowe) [inline components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ExternalEventAcceptanceTest.kt]: I think a constant would be better here rather than the magic number

Comment (ben-millar) [reply components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ExternalEventAcceptanceTest.kt]: That's a fair point, it's a little unclear what `500000L` is. I'll address this if I have to make any other changes to the PR.

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/RequestHandlerTestContext.kt [hunk 3]
```diff
@@ -57,5 +57,5 @@ class RequestHandlerTestContext<PAYLOAD>(val payload: PAYLOAD) {
         whenever(flowCheckpoint.holdingIdentity).thenReturn(holdingIdentity.toCorda())
     }
 
-    val flowEventContext = FlowEventContext(flowCheckpoint, flowEvent, payload, flowConfig, recordList, mdcProperties = emptyMap())
+    val flowEventContext = FlowEventContext(flowCheckpoint, flowEvent, payload, flowConfig, false, recordList, mdcProperties = emptyMap())
```
Comment (emilybowe) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/RequestHandlerTestContext.kt]: can you name the boolean parameter?

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt [hunk 4]
```diff
@@ -412,8 +415,10 @@ class ExternalEventManagerImplTest {
         assertEquals(payload.array(), record.value)
     }
 
-    @Test
-    fun `getEventToSend does not return an external event if the state is RETRY and the sendTimestamp is not surpassed`() {
+    @ParameterizedTest(name = "getEventToSend does not return an external event if the state is {0} and the sendTimestamp is not surpassed")
```
Comment (emilybowe) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt:418]: Should the state be {1} here ?

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt [hunk 5]
```diff
@@ -412,8 +415,10 @@ class ExternalEventManagerImplTest {
...
+    @EnumSource(names = ["RETRY", "OK"])
+    @Suppress("MaxLineLength")
+    fun `getEventToSend does not return an external event if the state is correct and the sendTimestamp is not surpassed`(stateType: ExternalEventStateType) {
```
Comment (emilybowe) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt]: Does 'correct' mean OK ?

**Review by emilybowe — COMMENTED** [review]
left some comments

**Review by emilybowe — DISMISSED** [review]
left some minor comments

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt (hunk 4 above)
Comment (ben-millar) [reply components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt:418]: No so this is actually just a string wildcard, `{0}` is replaced with the enum passed in by the parametrized decorator (either `RETRY` or `OK`) when the tests are built

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt (hunk 5 above)
Comment (ben-millar) [reply components/flow/flow-service/src/test/kotlin/net/corda/flow/external/events/impl/ExternalEventManagerImplTest.kt]: Correct here means any of the states that can be safely retried, but maybe I should reword this to `any non-error stage` for clarity

File: components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ExternalEventAcceptanceTest.kt [hunk 6]
```diff
@@ -253,6 +255,40 @@ class ExternalEventAcceptanceTest : FlowServiceTestBase() {
         }
     }
 
+    @Test
+    fun `Receiving an event resends the external event if status is OK but the retry window has been surpassed`() {
```
Comment (LWogan) [inline components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ExternalEventAcceptanceTest.kt:259]: Theres a file we use to write down all the test cases for acceptance tests as well to help identify coverage issues as theres so many across multiple tests. please add this test to it

flow-pipeline-acceptance-test-coverage.md

File: libs/flows/flow-api/src/main/kotlin/net/corda/flow/pipeline/events/FlowEventContext.kt [hunk 7]
```diff
@@ -15,6 +15,7 @@ import net.corda.messaging.api.records.Record
  * @param checkpoint The [FlowCheckpoint] of a flow that should be modified by the pipeline.
  * @param inputEvent The received [FlowEvent].
  * @param inputEventPayload The received [FlowEvent.payload].
+ * @param isRetryEvent Indicates whether we should clear the retry state in the flow checkpoint when this event is processed.
```
Comment (LWogan) [inline libs/flows/flow-api/src/main/kotlin/net/corda/flow/pipeline/events/FlowEventContext.kt]: I think we should use to flag to signal that we are processing a retry event. Some current (and possibly additional future) logic might want to hang off of its value. I think the kdoc should be updated to just say what the flag means which is that we are currently processing a retry event. 

I dont think we should set isRetryEvent to be false at any point in the pipeline while we are still processing it.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt [hunk 8]
```diff
@@ -130,10 +130,13 @@ class FlowGlobalPostProcessorImpl @Activate constructor(
...
-        // If we reach the post-processing step with a retry set we
-        // assume whatever the previous retry was it has now cleared
-        log.debug("The Flow was in a retry state that has now cleared.")
-        checkpoint.markRetrySuccess()
+        if (context.isRetryEvent) {
+            // If we reach the post-processing step with a retry set we
+            // assume whatever the previous retry was it has now cleared
+            log.debug("The Flow was in a retry state that has now cleared.")
+            checkpoint.markRetrySuccess()
+            context.isRetryEvent = false
```
Comment (LWogan) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt]: see comment below regarding setting this to be false

**Review by LWogan — COMMENTED** [review]
Mostly looks good to me. Sorry for the delay!

**Review by JamesHR3 — APPROVED** [review]
